src/main/java/com/google/vinet/data/MatchingAlgorithm.java

```diff
@@ -0,0 +1,160 @@
+   * availability times.
+   *
+   * <p>This method has a worst-case run time of O(N*M + E*sqrt(N+M)), where N is the number of
+   * volunteer time slots, M the number of isolate time slots, and E the number of edges between the
+   * two sets. The O(N*M) is due to the edge creation before running the algorithm.
+   *
+   * @param isolateTimeSlots   The set of all requested time slots for help
+   * @param volunteerTimeSlots The set of all time slots in which volunteers are available to help
+   * @return A set of matched time slots, where a volunteer was matched to a requested time slot
+   */
+  public static Set<IsolateTimeSlot> matchTimeSlots(
+          Set<IsolateTimeSlot> isolateTimeSlots, Set<VolunteerTimeSlot> volunteerTimeSlots) {
+
+    if (isolateTimeSlots == null || volunteerTimeSlots == null)
+      throw new IllegalArgumentException("Null argument!");
+
+    // Remove null nodes, if present
+    isolateTimeSlots.remove(null);
+    volunteerTimeSlots.remove(null);
+
+    addEdges(isolateTimeSlots, volunteerTimeSlots);
+    isolateTimeSlots.add(NIL_NODE);
+
+    while (breadthFirstSearch(isolateTimeSlots)) {
+      for (IsolateTimeSlot isolateTimeSlot : isolateTimeSlots) {
+        if (!isolateTimeSlot.isPaired()) {
+          depthFirstSearch(isolateTimeSlot);
+        }
+      }
+    }
+
+    return isolateTimeSlots.stream().filter(TimeSlot::isPaired).collect(Collectors.toSet());
+  }
+
+  /**
+   * Adds edges between the two sets of time slots, based on availability constraints.
+   *
+   * <p>Adds an edge if a volunteer time slot contains an isolate time slot.
+   *
+   * <p>This is an N*M operation, where N is the number of isolate time slots and M the number of
+   * volunteer time slots.
+   */
+  private static void addEdges(
+          Set<IsolateTimeSlot> isolateTimeSlots, Set<VolunteerTimeSlot> volunteerTimeSlots) {
+    for (VolunteerTimeSlot volunteerTimeSlot : volunteerTimeSlots) {
+      for (IsolateTimeSlot isolateTimeSlot : isolateTimeSlots) {
+        // TODO check if within geographic range
+        // TODO consider the case where volunteer time slot is longer than isolate's
+        // we could run the algorithm again with the remaining isolate slots and the chunks of
+        // volunteer time slots that were not assigned
+
+        // Remove time slot if any of its instants are null
+        if (volunteerTimeSlot.getStart() == null || volunteerTimeSlot.getEnd() == null) {
+          volunteerTimeSlots.remove(volunteerTimeSlot);
+          continue;
+        } else if (isolateTimeSlot.getStart() == null || isolateTimeSlot.getEnd() == null) {
+          isolateTimeSlots.remove(isolateTimeSlot);
+          continue;
+        }
+
+        if (volunteerTimeSlot.contains(isolateTimeSlot)) {
+          isolateTimeSlot.addNeighbour(volunteerTimeSlot);
+          volunteerTimeSlot.addNeighbour(isolateTimeSlot);
+        }
```

Comment:
No, because the Isolate time slot is actually a request for that exact amount of time, _at_ that exact moment in time. I have now added a function `hasMinimumOverlap` which check that, because it can be useful if we improve the algorithm to take into account isolate availability times too.

Reply:
Ya, cleaning this up a bit is fine.

Reply:
Ah I see. That's a rather odd design choice, no? I thought the isolate request is like, "I need 30 mins from volunteers within the window of 3pm to 7pm"... then you need to see if a volunteer and isolate has 30 min time overlap.

Reply:
Yes, however I don't know of an algorithm that would be able to solve that together with the other requirements in a resonable amount of time, hence the choice.